moby/moby
# cli: long-syntax support for `--volume` (deprecates `--mount` for `docker service`)

**- What I did**

This PR implements the new UX guideline concluded in https://github.com/docker/docker/issues/28527#issuecomment-262817805

> Mounts: deprecate --mount (maybe remove it from run immediately) and have --volume support the long syntax.

The long-syntax is compatible to `--mount` except that the flag name is `--volume`.

 * `docker run`:     supports both the short-syntax and the long-syntax
 * `docker service`: only supports the long-syntax

The long/short detection is done by using some regexp matching implemented in
`opts/volume.go`. (see `InferVolumeType()`)

This commit also introduces `--volume-add` and `--volume-rm` for `docker service`,
and hence deprecates `--mount-add` and `--mount-rm`.
Note that `--mount`, `--mount-add`, and `--mount-rm` are still available for
compatibility.

**- How I did it**

Please refer to `opts/volume.go`

```go
// InferVolumeType infers the VolumeType
func InferVolumeType(s string) VolumeType {
        // NOTE: we can't use `(,\\w+=\\w+)*` (originally suggested in #28527),
        // because `\w` lacks '/' and non-ASCII chars
        matchedLong, err := regexp.MatchString(".+=.+", s)
        if err != nil {
                // should not happen
                return InvalidVolume
        }
        if matchedLong {
                return LongVolume
        }
        if len(volumeSplitN(s, 2)) > 1 {
                return ShortSplitVolume
        }
        return ShortUnsplitVolume
}
```

**- How to verify it**

 * Make sure `docker run --volume type=volume,source=foo,destination=/bar` works
 * Make sure `docker run -v foo:/bar` works
 * Make sure `docker service create --volume type=volume,source=foo,destination=/bar` works
 * Make sure `docker service create --mount type=volume,source=foo,destination=/bar` works
 * Make sure `docker service create -v foo:/bar` fails with an error like `unsupported volumes (needs to be in long-syntax): [foo:/bar]`.


**- Description for the changelog**

cli: long-syntax support for `--volume` (deprecates `--mount` for `docker service`)

**- A picture of a cute animal (not mandatory but encouraged)**
![king_penguin_chick_at_salisbury_plain_ 5719980176](https://cloud.githubusercontent.com/assets/9248427/22240620/487d676c-e25f-11e6-8847-4712ba9dd724.jpg)



Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>
